src/python/pants/backend/python/check/pip_audit/rules.py

```diff
@@ -0,0 +1,169 @@
+
+import re
+from collections import defaultdict
+from dataclasses import dataclass
+from hashlib import sha256
+from typing import Mapping
+
+from pants.backend.python.check.pip_audit.skip_field import SkipPipAuditField
+from pants.backend.python.check.pip_audit.subsystem import PipAudit
+from pants.backend.python.subsystems.setup import PythonSetup
+from pants.backend.python.target_types import PythonRequirementResolveField, PythonRequirementsField
+from pants.backend.python.util_rules import pex_from_targets
+from pants.backend.python.util_rules.partition import ResolveName
+from pants.backend.python.util_rules.pex import PexRequest, PexResolveInfo, VenvPex, VenvPexProcess
+from pants.backend.python.util_rules.pex_from_targets import RequirementsPexRequest
+from pants.base.build_root import BuildRoot
+from pants.core.goals.check import CheckRequest, CheckResult, CheckResults
+from pants.engine.collection import Collection
+from pants.engine.fs import CreateDigest, Digest, FileContent
+from pants.engine.process import FallibleProcessResult, ProcessCacheScope
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import FrozenOrderedSet, OrderedSet
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class PipAuditFieldSet(FieldSet):
+    required_fields = (
+        PythonRequirementsField,
+        PythonRequirementResolveField,
+    )
+
+    resolve: PythonRequirementResolveField
+    sources: PythonRequirementsField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipPipAuditField).value
+
+
+@dataclass(frozen=True)
+class PipAuditPartition:
+    description: str
+    field_sets: FrozenOrderedSet[PipAuditFieldSet]
+
+
+class PipAuditPartitions(Collection[PipAuditPartition]):
+    pass
+
+
+class PipAuditRequest(CheckRequest):
+    field_set_type = PipAuditFieldSet
+    name = PipAudit.options_scope
+
+
+@rule
+async def pip_audit_partition(
+    partition: PipAuditPartition,
+    pip_audit: PipAudit,
+    build_root: BuildRoot,
+) -> CheckResult:
+    requirements_pex_get = Get(
+        VenvPex,
+        RequirementsPexRequest(
+            (fs.address for fs in partition.field_sets),
+        ),
+    )
+
+    pip_audit_pex_get = Get(
+        VenvPex,
+        PexRequest,
+        pip_audit.to_pex_request(),
+    )
+
+    requirements_pex, pip_audit_pex = await MultiGet(requirements_pex_get, pip_audit_pex_get)
+    requirements_pex_info = await Get(PexResolveInfo, VenvPex, requirements_pex)
+    requirements_str = "\n".join(
+        "{}=={}".format(dist_info.project_name, dist_info.version)
+        for dist_info in requirements_pex_info
+    )
+    requirements_digest = await Get(
+        Digest,
+        CreateDigest([FileContent("requirements.txt", requirements_str.encode())]),
```

Comment:
Not sure if this file will cause collisions in Pants' cache when multiple partitions are used or if the build root has a file with the same name?